# In-depth NGS Data Analysis Course

| Audience | Computational Skills | Prerequisites | Duration |
:----------|:----------|:----------|:----------|
| Biologists | Beginner/Intermediate | None | 12-day workshop (~78 hours of trainer-led time)|

This repository contains all course materials for the Summer 2017 version of HBC's In-depth NGS Data Analysis Course, a 12-day course run over 6 weeks.

> These materials are developed for a trainer-led workshop, but they are also amenable to self-guided learning.

The course is aimed at bench biologists who are interested in learning about NGS-based genomic analysis. The topics covered in-depth during this course are analysis of RNA-Seq and ChIP-Seq data, with an optional Variant Calling session. The sessions will also include functional analysis downstream of sequence data processing. During this course, participants will gain skills in the areas of:
 
* UNIX and basic shell scripting
* high-performance compute clusters, and 
* R for statistical analysis and data visualization. 

At the end of this course, participants can expect to have the expertise to independently run data analysis for sequencing experiments.

**No prior programming experience or command-line training is required.**

This repo contains the materials for the six sessions of the course. The six sessions are described below:

 * [**Session I**: Introduction to Unix / Orchestra and NGS Data Analysis](https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionI/README.html)
 * [**Session II**: RNA-Seq Part I](sessionII)
 * [**Session III**: RNA-Seq Part II](sessionIII)
 * [**Session IV**: RNA-Seq Part III and Related Technologies and Tools](sessionIV)
 * [**Session V**: ChIP-Seq](sessionV)
 * [**Session VI**: Variant Calling](sessionVI)

---

*NOTE: Additional materials are included in this repo but are not part of the main course.*

---

### Installation Requirements

Download and install the following programs:
>**NOTE:** If you are going to be using a **Mac laptop** please use the [instructions here](https://support.apple.com/en-us/HT201260) to determine your OS X version. If you find that your Mac is running a version older than OS X 10.6, please email us.

**Mac users:**
- [Java](https://www.java.com/en/download/)
- [Filezilla](https://filezilla-project.org/download.php?type=client)
- [Integrative Genomics Viewer (IGV)](https://www.broadinstitute.org/igv/?q=registration)
- [Sublime Text](http://www.sublimetext.com/)
- [Rstudio](http://www.rstudio.com/products/rstudio/download/) (pick the appropriate file for your operating system)
- [R](http://cran.r-project.org/bin/macosx/) (R-3.4.2 for Mac OS X 10.11 (El Capitan) and higher, R-3.3.3 for Mac OS X 10.9 (Mavericks) and R-3.1.3 for Mac OS X 10.6 (Snow Leopard), OS X 10.7 and OS X 10.8). 

**Windows users:**
- [GitBash](https://git-scm.com/download/win)
- [Java](https://www.java.com/en/download/)
- [Filezilla](https://filezilla-project.org/)
- [Integrative Genomics Viewer (IGV)](https://www.broadinstitute.org/igv/?q=registration)
- [Notepad++](http://notepad-plus-plus.org/) or [Sublime Text](http://www.sublimetext.com/)
- [Rstudio](http://www.rstudio.com/products/rstudio/download/) (pick the appropriate file for your operating system)
- [R](http://cran.r-project.org/bin/windows/base/)

*These materials have been developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*

* *Some materials used in these lessons were derived from work that is Copyright Â© Data Carpentry (http://datacarpentry.org/). 
All Data Carpentry instructional material is made available under the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0).*
