---
title: "Introduction to knitr"
author: "Michael J. Steinbaugh"
date: "2017-07-06"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

[knitr]: https://yihui.name/knitr/
[RMarkdown]: http://rmarkdown.rstudio.com/
[R]: https://www.r-project.org/
[RStudio]: https://www.rstudio.com/



[knitr][], developed by [Yihui Xie](https://yihui.name), is an R package designed for report generation within [RStudio][]. It enables dynamic generation of multiple file formats from an [RMarkdown][] file, including HTML and PDF documents.

As RMarkdown grows as an acceptable [reproducible manuscript](https://elifesciences.org/labs/cad57bcf/composing-reproducible-manuscripts-using-r-markdown) format, using [knitr][] to generate a report summary is becoming common practice.

Document creation is now integrated into [RStudio][], and can be accessed using the GUI (show image of knit button) or two command line options:


# chunks

In RMarkdown, programming chunks are defined by three backticks ("```") then then programming language, followed by a *unique* chunk label, and customization options.


## Code languages

Recently [RStudio][] has added support for additional [code languages](http://rmarkdown.rstudio.com/lesson-5.html) besides R inside code including bash and python:

- Python
- SQL
- Bash
- Rcpp
- Stan
- JavaScript
- CSS

This is still in development but allows for integration of multiple programming languages into a single workflow, and is shaping up to be a very powerful tool for NGS analysis in the future.


## Global options

```{r global, eval=FALSE}
library(knitr)
opts_chunk$set(
    autodep = TRUE,
    cache = TRUE,
    cache.lazy = TRUE,
    dev = c("png", "pdf", "svg"),
    error = TRUE,
    fig.height = 6,
    fig.retina = 2,
    fig.width = 6,
    highlight = TRUE,
    message = FALSE,
    prompt = TRUE,
    tidy = TRUE,
    warning = FALSE)
```


## Per chunk options

Walk through some examples of useful per chunk options:

- `echo = FALSE`
- `eval = FALSE`
- `include = FALSE`
- `message = FALSE`
- `results = "asis"`
- `warning = FALSE`

Resize an image:

- `fig.height = 6`
- `fig.width = 4`



# `setup` chunk

The `setup` chunk is a special knitr chunk that should be placed at the start of the document. We recommend storing all `library()` loads required for the script and other `load()`/`data()` requests for external files here. In our RMarkdown templates, such as the bcbioRnaseq [differential expression template](https://github.com/hbc/bcbioRnaseq/blob/master/inst/rmarkdown/templates/differential_expression/skeleton/skeleton.Rmd), we store all the user-defined parameters in the `setup` chunk that are required for successful knitting.

> {r setup, include=FALSE}
> knitr::opts_chunk$set(echo = TRUE)


## knitr::knit

```{r knit}
help("knit", "knitr")
```

When you `knit()` a document, by default this will generate an HTML report. If you would prefer a different document format, this can be specified in the YAML header with the `output:` parameter.

*refer back to RMarkdown table*


## rmarkdown::render

```{r render}
help("render", "rmarkdown")
```

Mention that `render()` allows for output of multiple document formats.

`_output.yaml` file from bcbioRnaseq [link](https://github.com/hbc/bcbioRnaseq/blob/master/docs/downloads/_output.yaml):

    rmarkdown::html_document:
        code_folding: hide
        df_print: kable
        highlight: pygments
        number_sections: false
        toc: true
    rmarkdown::pdf_document:
        number_sections: false
        toc: true
        toc_depth: 1


Note on pandoc? PDF rendering can fail on some systems, may require Homebrew install.



# kable

Section about stylized tables using `kable`.

```{r}
help("kable", "knitr")
```

Alignment parameters from basejump scripts, previous consults [need to add].

Alternatives:

- xtable
- pandoc table



# Working directory behavior

**Note**: [knitr][] redefines the working directory of an RMarkdown file in a manner that can be confusing. If you're working in [RStudio][] with an [RMarkdown][] file that is not at the same location as the current R working directory (`getwd()`), you can run into problems with broken file paths. Let's say I have [RStudio][] open without a project loaded. My working directory is set to `~/Users/mike`. Now, if I load an RMarkdown file from my desktop at `~/Users/mike/Desktop`, [knitr][] will set the working directory within chunks to be relative to my desktop. We advise against coding paths in a script to only work with [knitr][] and not base [R][].

A simple way to resolve this issue is by creating an R project for the analysis, and saving all RMarkdown files at the top level, to avoid running into unexpected problems related to this behavior.



# Convert an R script to an RMarkdown knit report

Let's convert our Mov10 DE analysis script into an RMarkdown report:

```{r download_script}
download.file(
    file.path("https://raw.githubusercontent.com",
              "hbctraining",
              "In-depth-NGS-Data-Analysis-Course",
              "may2017",
              "sessionVI",
              "scripts",
              "de_script_to_knit.R"),
    destfile = "de_script_to_knit.R")
```

Open this script as a new tab in RStudio. Let's create a new RMarkdown called `de_analysis.Rmd`. You can do this quickly in RStudio by going to `File -> New File -> R Markdown...`. Customize the RMarkdown YAML header and create chunks for each code block.
