---
Audience: Biologists
Computational Skills: Beginner/Intermediate
Prerequisites: None
Duration: 2-day workshop (~13 hours of trainer-led time)
---

## Session I: Introduction to Unix / Orchestra and NGS Data Analysis

### Description

Session I will start with an overview of bioinformatics and next-generation sequencing (NGS), followed by an interactive introduction to Unix and Orchestra (the high performance computational cluster maintained by HMS Research Computing). The session will wrap up with an in-depth introduction to sequencing technology, library preparation and QC protocols for NGS data.

**Day 1:** We will start the course by laying out the path for the NGS Analysis course and providing a general overview of bioinformatics and NGS data. The remainder of the day will be dedicated to learning Unix, including navigating filesystems and performing basic operations.

**Day 2:** The second day will start with an introduction to computational clusters and O2, which will provide a thorough description of what the cluster environment is and how we benefit from using it. Following the introduction to O2, we will discuss best practices in Research Data Management (RDM) and spend some time implementing some of these to organize our project before jumping into the workflow. We will then introduce how Illumina sequencing works and what goes into preparing libraries for sequencing transcriptomics experiments. Finally, we will learn how to assess and interpret the quality of sequencing (FASTQ) files, discuss different error profiles you can expect to see with Illumina sequencers, and the main sources of errors.

> These materials were developed for a trainer-led workshop, but are also amenable to self-guided learning.

### Learning Objectives

1.	Understand the necessity for, and use of, the command line interface (bash) and HPC for analyzing high-throughput sequencing data.
2.	Understand best practices for designing an RNA-seq experiment and analysis the resulting data.


### Contents

| Lessons            | Estimated Duration |
|:------------------------|:----------|
|[Introduction to the NGS Data Analysis Course](lectures/) | 40 min |
|[Overview of bioinformatics](lectures/) | 35 min |
|[Introduction to Unix shell and O2](lectures/) | 15 min |
|[Unix shell - Files and directories](https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/01_the_filesystem.html) | 130 min |
|[Unix shell - Searching and redirection](https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/02_searching_files.html) | 45 min |
|[Unix shell - Introduction to vim](https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/03_vim.html) | 35 min |
|[Unix shell - Loops, variables and shell scripts](https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/04_loops_and_scripts.html) | 75 min |
|[Unix shell - Permissions and environment variables](https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/05_permissions_and_environment_variables.html) | 60 min |
|[Introduction to High-Performance Computing
(HPC)](lectures/) | 50 min |
|[Project organization and data management](https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/06_data_organization.html) | 40 min |
|[Research data management concepts](lectures/) | 30 min |
|[Illumina sequencing technology +
RNA-seq design & analysis](lectures/) | 60 min |
|[Library preparation](lectures/) | 90 min |
|[FastQ format and Illumina error profiles](lectures/) | 40 min |
|[QC: Quality assessment](lessons/10_salmon.md) | 55 min |
