## Session III: RNA-Seq Part II

### Description

Session III starts with an interactive lesson in R, wherein we cover advanced topics including how to match up and reorder rows and columns based on their names. We will then use various packages in R to perform the differential gene expression (DGE) analysis on the count matrix created in Session II, to generate a list of differentially expressed (DE) genes. In addition to performing this analysis, we will be talking about the QC steps and statistics that are involved in the DGE analysis. The list(s) of DE genes will then be used as input to various tools to graphically represent the results of this analysis. Finally, this session will cover gene annotation and functional enrichment analysis using R packages such as BiomaRt clusterProfiler and gProfileR.   

> These materials were developed for a trainer-led workshop, but are also amenable to self-guided learning.




### Contents

| Lessons            | Estimated Duration |
|:------------------------|:----------|
| [Matching and reordering ](lessons/01_advR-matching.md) | |
| [DGE setup and overview](lessons/02_DGE_setup_and_overview.md) | |
| [Count normalization ](lessons/03_DGE_count_normalization.md) | |
| [Exploratory data analysis](lessons/04_DGE_QC_analysis.md) | |
| [DESeq2 workflow overview](lessons/05_DGE_DESeq2_analysis.md) | |
| [DGE analysis: Pairiwse comparisons (Wald test)](lessons/06_DGE_DESeq2_analysis2.md) | |
| [DGE results summarization and visualization ](lessons/07_DGE_visualizing_results.md) | |
| [Workflow summarization ](lessons/08_DGE_summarizing_workflow.md) | |
| [DGE analysis: Complex experimental designs (LRT) ](lessons/09_DGE_LRT.md) | |
| [Ensembl/Biomart ](lessons/10_Ensembl_biomart.md) | |
| [Functional Enrichment analysis ](lessons/11_functional_analysis.md) | |
