## Session II: Introduction to Unix / O2 and NGS Data Analysis

### Description

Session II will start with a discussion of common file formats encountered during NGS analyses. We will then tackle RNA-Seq read alignment with STAR, assess the alignment quality, visualize alignments, and count the number of reads associated with genes with featureCounts. We will finish the session by introducing R, which we need to start with differential expression analysis of our count data.

### Contents

| Lessons            | Estimated Duration |
|:------------------------|:----------|
| [File formats in NGS analysis](lectures/) |
| [Sequence alignment tools and theory](lectures/) |
| [Alignment and counting](https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/08_rnaseq_workflow.html) |
| [Assessing alignment quality and visualization with IGV]() |
| [Generating a count matrix]() |
| [Intro to R and RStudio]() |
| [R syntax and data structure]() |
| [Intro to R: Functions and arguments]() |
| [Intro to R: Data wrangling vectors and factors]() |
| [Intro to R: Data wrangling matrices, data frames and lists]() |
| [Intro to R: Nested functions]() |
### Practical exercises
After completion of the session, practice of concepts can be explored with [these exercises]().
