## Session V: ChIP-Seq

### Description

Session V will focus on ChIP-Seq analysis. We will start with read quality control and alignment, then discuss various tools for peak calling, ChIP-Seq QC and visualization for ChIP-Seq analysis. We will talk about the IDR framework when handling replicates with ChIP-seq data and walk-through running IDR on true replicates. Since we have samples from two different IPs we will use DiffBind to demonstrate how to perform differential binding analyses. Finally, we will perform functional analysis of the peak calls, ending our analysis of the ChIP-Seq data. We will end the course by going over a few additional helpful tools. 

### Contents
 
| Lessons            | Estimated Duration |
|:------------------------|:----------:|
|[Introduction to ChIP-Seq](lectures/)| 60 min |
|[Set-up and project organization](lessons/01_Intro_chipseq_and_setup.md)| 20 min |
|[QC and alignment](lessons/02_QC_and_alignment.md)| 60 min |
|[Peak calling with MACS](lessons/03_peak_calling_macs.md)| 90 min |
|[ChIP QC: strand cross-correlation](lessons/04_QC_cross_correlation.md)| 70 min |
|[ChIP QC: sample level QC with ChIPQC](lessons/05_QC_quality_metrics.md)| 60 min |
|[Assessing replicates: bedtools and IDR](lessons/06_handling-replicates.md)| 90 min |
|[ChipSeeker: annotation and functional analysis](lessons/07_ChIPseeker_functional_analysis.md)| 30 min |
|[Web-based functional analysis](lessons/08_functional_analysis.md)| 90 min |
|[Differential peak enrichment with DiffBind](lessons/09_differential_peaks.md)| 45 min |
|[Visualization of differential peak enrichment using IGV](lessons/10_QC_visualization.md)| 45 min |
|[Integrating RNAseq and ChIPseq](lessons/integrating_rna-seq_and_chip-seq.md)| 20 min |
|[Orchestra mounting](lessons/orchestra_mounting.md)| 60 min |
