# Schedule for Session V


## Day 1

| Time            |   Topic  | Instructor |
|:------------------------:|:----------:|:--------:|
|09:00 - 10:00 | [Introduction to ChIP-seq]() | Dr. Shannan Ho Sui |
|10:00 - 10:30 | [Setting up for ChIP-seq analysis]() | Mary |
|10:30 - 10:45 | Break | |
|10:45 - 11:15 | [Best Practices in Research Data Management]() | Radhika |
|11:15 - 12:00 | [Sequencing data QC using FastQC](https://hbctraining.github.io/Intro-to-ChIPseq/lessons/02_QC_FASTQC.html) | Mary |
|12:00 - 13:00 | Lunch | |
|13:00 - 14:00 | [Alignment and filtering of reads](https://hbctraining.github.io/Intro-to-ChIPseq/lessons/03_align_and_filtering.html) | Mary |
|14:00 - 15:20 | [Automating generation of alignment files](https://hbctraining.github.io/Intro-to-ChIPseq/lessons/04_automation.html) | Radhika |
|15:20 - 15:30 | Break | |
|15:30 - 15:50 | [ChIP-seq File Formats](https://hbctraining.github.io/Intro-to-ChIPseq/lectures/Workflows_and_fileformats.pdf) | Radhika |
|15:50 - 17:00 | [Peak calling with MACS2](https://hbctraining.github.io/Intro-to-ChIPseq/lessons/05_peak_calling_macs.html) | Meeta |


## Day 3

| Time            |  Topic  | Instructor |
|:------------------------:|:----------:|:--------:|
|9:00 - 9:20 | [Peak calling with MACS2 (contd.)](https://hbctraining.github.io/Intro-to-ChIPseq/lessons/05_peak_calling_macs.html) | Meeta |
|9:20 - 10:20 | [Assessing ChIP quality using cross correlation](https://hbctraining.github.io/Intro-to-ChIPseq/lessons/06_QC_cross_correlation.html) | Kayleigh |
|10:20 - 10:35 | Break | |
|10:35 - 11:35| [Assessing Peak calls and ChIP quality using ChIPQC](https://hbctraining.github.io/Intro-to-ChIPseq/lessons/07_QC_quality_metrics.html) | Meeta |
|11:35 - 12:00 | [Handling Replicates with Bedtools](https://hbctraining.github.io/Intro-to-ChIPseq/lessons/08_handling-replicates-bedtools.html) | Radhika |
|12:00 - 13:00 | Lunch | |
|13:00 - 13:50 | [Handling Replicates with IDR](https://hbctraining.github.io/Intro-to-ChIPseq/lessons/09_handling-replicates-idr.html) | Radhika |
|13:50 - 15:10 | [Visualization and exploration of ChIP-seq data](https://hbctraining.github.io/Intro-to-ChIPseq/lessons/10_data_visualization.html) | Meeta |
|15:10 - 15:25 | Break | |
|15:25 - 15:50 | [Qualitative assessment using IGV](https://hbctraining.github.io/Intro-to-ChIPseq/lessons/11_qualitative_assessment_IGV.html) | Radhika |
|15:50 - 16:20 | [Functional analysis](https://hbctraining.github.io/Intro-to-ChIPseq/lessons/12_functional_analysis.html) | Kayleigh |
|16:20 - 16:40 | [Overview of ChIP-seq workflow](https://hbctraining.github.io/Intro-to-ChIPseq/lectures/ChIP-seq_troubleshooting.pdf) | Meeta |
|16:40 - 17:00 | [Wrap-up and Survey](https://hbctraining.github.io/Intro-to-ChIPseq/lectures/Wrap-up.pdf) | Radhika |

**Dataset:** [Introduction to shell dataset](https://www.dropbox.com/s/3lua2h1oo18gbug/unix_lesson.tar.gz?dl=1)
